src/ray/core_worker/core_worker.cc

```diff
@@ -12,23 +12,37 @@ CoreWorker::CoreWorker(const enum WorkerType worker_type,
       store_socket_(store_socket),
       raylet_socket_(raylet_socket),
       worker_context_(worker_type, job_id),
-      raylet_client_(raylet_socket_,
-                     ClientID::FromBinary(worker_context_.GetWorkerID().Binary()),
-                     (worker_type_ == ray::WorkerType::WORKER),
-                     worker_context_.GetCurrentJobID(), ToTaskLanguage(language_)),
       task_interface_(*this),
       object_interface_(*this),
       task_execution_interface_(*this) {
-  // TODO(zhijunfu): currently RayletClient would crash in its constructor if it cannot
-  // connect to Raylet after a number of retries, this needs to be changed
-  // so that the worker (java/python .etc) can retrieve and handle the error
-  // instead of crashing.
   auto status = store_client_.Connect(store_socket_);
   if (!status.ok()) {
     RAY_LOG(ERROR) << "Connecting plasma store failed when trying to construct"
                    << " core worker: " << status.message();
     throw std::runtime_error(status.message());
   }
+
+  // NOTE(zhijunfu): For non-driver worker, the initialization of
+  // raylet client is delayed to when `TaskExecutionInterface::Run()`
+  // is called, as it is until that time the port for the worker
+  // rpc server can be determined, and this information needs to be
+  // included when worker registers to raylet.
+  if (worker_type_ == WorkerType::DRIVER) {
+    InitializeRayletClient(0);
+  }
+}
+
+void CoreWorker::InitializeRayletClient(int server_port) {
+  if (raylet_client_ == nullptr) {
```

Comment:
I would change this to a check that `raylet_client_` is not already initialized.

Reply:
this will be removed as we separate `driver` and `worker`.